dirty_cat/deduplicate.py

```diff
@@ -0,0 +1,177 @@
+    unrolled_corrections = translation_table[data]
+    if return_translation_table:
+        return unrolled_corrections, translation_table
+    else:
+        return unrolled_corrections
+
+
+def guess_clusters(Z: np.ndarray, distance_mat: np.ndarray) -> int:
+    """Finds the number of clusters that maximize the silhouette score
+    when clustering `distance_mat`.
+
+    Parameters
+    ----------
+    Z : np.ndarray
+        hierarchical linkage matrix, specifies which clusters to merge.
+    distance_mat : np.ndarray
+        distance matrix either in square or condensed form.
+
+    Returns
+    -------
+    int
+        number of clusters that maximize the silhouette score.
+    """
+    max_clusters = distance_mat.shape[0]
+    n_clusters = np.arange(2, max_clusters)
+    # silhouette score needs a redundant distance matrix
+    redundant_dist = squareform(distance_mat)
+    silhouette_scores = []
+    for n_clust in n_clusters:
+        labels = fcluster(Z, n_clust, criterion="maxclust")
+        silhouette_avg = silhouette_score(redundant_dist, labels, metric="precomputed")
+        silhouette_scores.append(silhouette_avg)
+    return n_clusters[np.argmax(silhouette_scores)]
+
+
+def create_spelling_correction(
+    unique_words: Union[Sequence[str], np.ndarray],
+    counts: Union[Sequence[int], np.ndarray],
+    clusters: Sequence[int],
+) -> pd.Series:
+    """Creates a pandas Series that map each cluster member to the most
+    frequent cluster member. The assumption is that the most common spelling
+    is the correct one.
+
+    Parameters
+    ----------
+    unique_words : Union[Sequence[str], np.ndarray]
+        A sequence or array of unique words in the original data.
+    counts : Union[Sequence[int], np.ndarray]
+        A sequence or array of counts of how often each unique word appears in
+        the original data.
+    count_series : pd.Series
+        A series with unique words (in the original data) as indices and number
+        of occurrences of each word in the original data as values.
+    clusters : Sequence[int]
+        A sequence of ints, indicating cluster membership of each unique word
+        in `count_series`.
+
+    Returns
+    -------
+    pd.Series
+        Series with unique (original) words as indices and (estimated)
+        corrected spelling of each word as values.
+    """
+    count_series = pd.Series(counts, index=unique_words)
+    original_spelling: List[str] = []
+    corrected_spelling: List[str] = []
+    for cluster in np.unique(clusters):
+        sorted_spellings = (
+            count_series.loc[clusters == cluster]
+            .sort_values(ascending=False)
+            .index.values
+        )
+        original_spelling.extend(sorted_spellings.tolist())
+        # assumes spelling that occurs most frequently in cluster is correct
+        corrected_spelling.extend(
+            np.repeat(sorted_spellings[0], sorted_spellings.shape[0])
```

Comment:
Don't use np.repeat as you need a list. Use multiplication on a list:
n * [1, ], this is better because you need a list at the end of the day, rather than a numpy array